src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -114,23 +112,43 @@ class LibraryService (protected val argUserInfo: UserInfo,
           case Failure(e) =>
             Future(RequestCompleteWithErrorReport(BadRequest, BadRequest.defaultMessage, e))
           case Success(x) => {
-            rawlsDAO.getWorkspace(ns, name) map  { workspaceResponse =>
-              // this is technically vulnerable to a race condition in which the workspace attributes have changed
-              // between the time we retrieved them and here, where we update them.
+            rawlsDAO.getWorkspace(ns, name) flatMap { workspaceResponse =>
               val allOperations = generateAttributeOperations(workspaceResponse.workspace.attributes, userAttrs,
                 k => k.namespace == AttributeName.libraryNamespace && k.name != LibraryService.publishedFlag.name)
-              RequestComplete(patchWorkspace(ns, name, allOperations, isPublished(workspaceResponse)))
+              // this is technically vulnerable to a race condition in which the workspace attributes have changed
+              // between the time we retrieved them and here, where we update them.
+              if (userAttrs.contains(discoverableWSAttribute)) {
+                if (userAttrs.size == 1)
+                  withDiscoverabilityModifyPermissions(workspaceResponse) {
+                    internalPatchWorkspaceAndRepublish(workspaceResponse.accessLevel, ns, name, allOperations, isPublished(workspaceResponse)) map (RequestComplete(_))
+                  }
+                else
+                  withDiscoverabilityModifyPermissions(workspaceResponse){
+                    withModifyPermissions(workspaceResponse){
+                      internalPatchWorkspaceAndRepublish(workspaceResponse.accessLevel, ns, name, allOperations, isPublished(workspaceResponse)) map (RequestComplete(_))
+                    }
+                  }
+              }
+              else {
+                withModifyPermissions(workspaceResponse){
+                  internalPatchWorkspaceAndRepublish(workspaceResponse.accessLevel, ns, name, allOperations, isPublished(workspaceResponse)) map (RequestComplete(_))
+                }
+              }
```

Comment:
What if you rewrote it like this?
```
              val doUpdate:Boolean = userAttrs match {
                case x if x.contains(discoverableWSAttribute) && x.size == 1 => canModifyDiscoverability(workspaceResponse)
                case x if x.contains(discoverableWSAttribute) => canModifyDiscoverability(workspaceResponse) && canModify(workspaceResponse)
                case _ => canModify(workspaceResponse)
              }

              if (doUpdate)
                internalPatchWorkspaceAndRepublish(workspaceResponse.accessLevel, ns, name, allOperations, isPublished(workspaceResponse)) map (RequestComplete(_))
              else
                throw FireCloudException("there was a problem") // not sure what to do here
```

trick is using the can* permission methods instead of the with* methods ...